### PR TITLE
RAND: Moving seeding out of the FIPS provider.

### PR DESCRIPTION
**Pull request by paulidale**
Due to a rule change by NIST about the validation requirements for operating system sources, it is necessary to move the seeding out of the FIPS provider and rely on a pull model instead.

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: crypto/evp/evp_rand.c
Comment (levitte) [inline crypto/evp/evp_rand.c:514]: ... and you still won't go for `OSSL_RAND`? :wink:

Comment (paulidale) [reply crypto/evp/evp_rand.c:514]: This would be part of a larger discussion about API naming 😞

**Comment by paulidale** [comment]
This needs a rethink.  A provided seed source isn't possible inside the FIPS provider.  An alternate method of getting seed material in is required.

**Comment by paulidale** [comment]
In addition to the seed callbacks, nonce ones are also required.

Also made the test RNG not a DRBG.  It isn't one and pretending to be one hid things that are better unhid.

**Comment by paulidale** [comment]
Assuming the CIs are happy, should be good for a reviewing.

**Comment by paulidale** [comment]
Ping for review.

**Comment by paulidale** [comment]
Re-ping for review.  @openssl/committers ?

File: providers/implementations/rands/drbg.c
Comment (slontis) [inline providers/implementations/rands/drbg.c:121]: do you always want r == 0? That is what it currently does..

File: providers/implementations/rands/drbg.c
Comment (slontis) [inline providers/implementations/rands/drbg.c:324]: should this be a clear_free?

Comment (paulidale) [reply providers/implementations/rands/drbg.c:324]: The get failed, I don't see that it needs to be cleared in this case.

File: providers/implementations/rands/drbg.c
Comment (paulidale) [reply providers/implementations/rands/drbg.c:121]: No, it doesn't 😁

**Comment by paulidale** [comment]
ping for review?

**Comment by paulidale** [comment]
Still awaiting @openssl/committers review...

File: providers/implementations/rands/crngt.c
Comment (mattcaswell) [inline providers/implementations/rands/crngt.c:126]: I don't understand this line.

File: providers/implementations/rands/drbg.c
Comment (mattcaswell) [inline providers/implementations/rands/drbg.c:202]: `r == 0`

**Review by mattcaswell — CHANGES_REQUESTED** [review]
Review comments so far...

File: providers/implementations/rands/crngt.c
Comment (paulidale) [reply providers/implementations/rands/crngt.c:126]: We don't currently support entropy sources that provide less than full entropy.  Once we do support such devices, we'll need a scaling step and it needs to go here.  I.e. this is a placeholder for such a step.  I can replace this line with a better comment if that would help.

File: doc/man7/provider-base.pod
Comment (paulidale) [reply doc/man7/provider-base.pod:70]: They don't need it currently, but including it is prudent.  I'll add.

File: doc/man7/provider-base.pod
Comment (paulidale) [reply doc/man7/provider-base.pod:220]: None of the other core functions include anything like this, I've raised an issue to improve them: #13358

File: providers/implementations/rands/crngt.c
Comment (paulidale) [reply providers/implementations/rands/crngt.c:126]: Added a comment and removed this line.

File: doc/man7/provider-base.pod
Comment (paulidale) [reply doc/man7/provider-base.pod]: Added

**Comment by paulidale** [comment]
Review comments addressed.

**Comment by paulidale** [comment]
Ping for rereview

**Comment by paulidale** [comment]
Re-ping for review.

File: providers/implementations/rands/crngt.c
Comment (mattcaswell) [inline providers/implementations/rands/crngt.c:161]: Doesn't `crngt_get_entropy` always populate the supplied buffer with `CRNGT_BUFSIZ` bytes? If so isn't this going to overrun the buffer on the last iteration if `t < CRNGT_BUFSIZ`?

File: providers/implementations/rands/drbg.c
Comment (mattcaswell) [inline providers/implementations/rands/drbg.c:185]: if locking if? This doesn't quite make sense

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: crypto/params.c
Comment (paulidale) [reply crypto/params.c:427]: It's truncating the low order bits of INT64_MAX so it fits into a double without loss of accuracy.

File: crypto/params.c
Comment (paulidale) [reply crypto/params.c:23]: It's both.  The mantissa field in an IEEE double is 52 bits in size.  The 53rd bit is implied since it will always be a _1_ after normalisation.

With the uptake of the IEEE standard, double should always be 4 or 8 bytes.  I've not seen a platform where IEEE wasn't used for decades (Vax has it's own floating point formats but double was still 8 bytes worked).

A check in `sanitytest.c` might be sensible, not for this PR though.  There are assumptions about four and eight byte integers and these aren't checked.

File: crypto/params.c
Comment (paulidale) [reply crypto/params.c:484]: Yes, normalisation is performed to make the 53rd bit a _1_.  Without this, the number wouldn't fit.

Comment (paulidale) [reply crypto/params.c:484]: This change shouldn't be here.

File: crypto/params.c
Comment (paulidale) [reply crypto/params.c:427]: This change shouldn't be here.

File: crypto/params.c
Comment (paulidale) [reply crypto/params.c:23]: This change shouldn't be here.

**Comment by paulidale** [comment]
What's confusing, is that some of the review comments are about a commit that shouldn't be in this PR.

File: providers/implementations/rands/drbg.c
Comment (paulidale) [reply providers/implementations/rands/drbg.c:185]: Reworded.

File: providers/implementations/rands/crngt.c
Comment (paulidale) [reply providers/implementations/rands/crngt.c:161]: Great catch.  Yes there is a risk here.  I don't think it's currently possible since everything is a multiple of CRNGT_BUFSIZ but better to make the code safe.

**Comment by paulidale** [comment]
Could someone with access to an ARM64 system let me know what's going on with the failed test?
It passes  here under ASAN.

File: providers/implementations/rands/crngt.c
Comment (mattcaswell) [inline providers/implementations/rands/crngt.c:162]: Don't we still need to do this if we used the `buf` in the last iteration of the loop above?

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: providers/implementations/rands/crngt.c
Comment (paulidale) [reply providers/implementations/rands/crngt.c:162]: Yes, we do need this.  `buf` came back to avoid the overflow.
I've added it.

**Review by mattcaswell — APPROVED** [review]

File: crypto/rand/prov_seed.c
Comment (mspncp) [inline crypto/rand/prov_seed.c:64]: remove [cat comment](https://www.reddit.com/r/ProgrammerHumor/comments/8w54mx/code_comments_be_like)?

**Review by mspncp — CHANGES_REQUESTED** [review]
Ok, here's a first last-minute pass over your PR, sorry for not responding earlier. I can't really say that I understood all details of the provider boilerplate code, but I have a lot of questions.

File: crypto/err/openssl.txt
Comment (mspncp) [reply crypto/err/openssl.txt:2968]: Ok, all-clear. It turns out that I already did the split in PR #12941, which was merged to master in commit b0614f0ae3c33182ad184dc82056b22aebf42956.

Comment (mspncp) [reply crypto/err/openssl.txt:2968]: Ok, but one thing remains to do: please rename your `RESEED_PROP_CTR` to `RESEED_COUNTER`.

File: providers/common/provider_err.c
Comment (mspncp) [inline providers/common/provider_err.c:156]: If I as a user would be confronted with the error message "parent cannot generate", I would be quite confused.

File: crypto/err/openssl.txt
Comment (paulidale) [reply crypto/err/openssl.txt:2968]: Okay, done.

Comment (mspncp) [reply crypto/err/openssl.txt:2968]: At 7:24 AM?

File: doc/internal/man3/ossl_rand_get_entropy.pod
Comment (paulidale) [reply doc/internal/man3/ossl_rand_get_entropy.pod:17]: The entropy argument is from an earlier form.  I've added more description of the salt parameter.

File: providers/common/provider_seeding.c
Comment (paulidale) [reply providers/common/provider_seeding.c:16]: These are only set by core at load time and always to the same values.
The same pattern is used elsewhere.  e.g. bio_prov.c

Moving these statics variables into each provider's structure isn't relevant to this PR, but might be something the future.

File: providers/common/provider_err.c
Comment (paulidale) [reply providers/common/provider_err.c:156]: Added "random numbers" to the message.

**Comment by paulidale** [comment]
Feedback addressed

File: doc/internal/man3/ossl_rand_get_entropy.pod
Comment (mspncp) [inline doc/internal/man3/ossl_rand_get_entropy.pod:40]: nit: contrains -> contains

File: doc/man7/provider-base.pod
Comment (mspncp) [inline doc/man7/provider-base.pod:228]: nit: seed material -> seeding material

**Review by mspncp — APPROVED** [review]
Ok, looks like all my comments have been addressed. I'm not going to dig further, for the rest I rely on @mattcaswell's review.

Just two nits, which are included in this approval.

**Comment by paulidale** [comment]
There were merge conflicts from #13417 which caused bit of pain.  They were function renaming issues, so no code changes.  I don't feel that reapproval is necessary however, I'll wait for the CIs before merging, just to be certain.

**Comment by paulidale** [comment]
The no-deprecated builds are failing with:
```
ssl/ssl_lib.c: In function 'SSL_set0_tmp_dh_pkey':

ssl/ssl_lib.c:5963:9: error: implicit declaration of function 'SSLerr'; did you mean 'stderr'? [-Werror=implicit-function-declaration]

         SSLerr(0, SSL_R_DH_KEY_TOO_SMALL);

         ^~~~~~
```
which isn't relevant to the changes here.

**Comment by paulidale** [comment]
Merged to master.  Thanks for the reviews, feedback and suggestions.

**Pull request closed without merging**
